api/app/auto_spatial_advisory/snow.py

```diff
@@ -0,0 +1,117 @@
+    # it's S3 virtual file system driver to read the file.
+    # https://gdal.org/user/virtual_file_systems.html
+    snow_key = f"/vsis3/{bucket}/snow_coverage/{for_date.strftime('%Y-%m-%d')}/clipped_snow_coverage_{for_date.strftime('%Y-%m-%d')}_epsg4326.tif"
+    snow_mask_warped_output_path = os.path.join(temp_dir, SNOW_COVERAGE_WARPED_NAME)
+    # Perform reprojection to Lambert Conformal Conic to match HFI data, crop extent and resample to 2km x 2km pixels
+    gdal.Warp(snow_mask_warped_output_path, snow_key, dstSRS=source_projection,
+              outputBounds=extent, xRes=x_res, yRes=y_res, resampleAlg=gdal.GRA_NearestNeighbour)
+    source = None
+    return snow_mask_warped_output_path
+    
+def apply_snow_mask_to_hfi(hfi_path: str, snow_mask_path: str, temp_dir: str):
+    # Open the hfi data
+    source_tiff = gdal.Open(hfi_path, gdal.GA_ReadOnly)
+    source_band = source_tiff.GetRasterBand(1)
+    source_data = source_band.ReadAsArray()
+
+    # Open the snow mask
+    snow_tiff = gdal.Open(snow_mask_path, gdal.GA_ReadOnly)
+    snow_band = snow_tiff.GetRasterBand(1)
+    snow_data = snow_band.ReadAsArray()
+
+    # The snow mask tiff has values of 0 or 1 where 0 represents areas covered by snow and 1 represents
+    # snow free areas. Multiply the rasters to apply the mask.
+    masked_data = np.multiply(source_data, snow_data)
+    masked_hfi_path = os.path.join(temp_dir, MASKED_HFI_PATH_NAME)
+    output_driver = gdal.GetDriverByName("GTiff")
+    # Create an object with the same dimensions as the hfi input
+    masked_hfi_tiff = output_driver.Create(masked_hfi_path, xsize=source_band.XSize,
+                                       ysize=source_band.YSize, bands=1, eType=gdal.GDT_Byte)
+    # Set the geotransform and projection to the same as the input.
+    masked_hfi_tiff.SetGeoTransform(source_tiff.GetGeoTransform())
+    masked_hfi_tiff.SetProjection(source_tiff.GetProjection())
+
+    # Write the classified data to the band.
+    masked_hfi_tiff_band = masked_hfi_tiff.GetRasterBand(1)
+    masked_hfi_tiff_band.SetNoDataValue(0)
+    masked_hfi_tiff_band.WriteArray(masked_data)
+
+    # Explicit delete to make sure underlying resources are cleared up!
+    source_data = None
+    source_band = None
+    source_tiff = None
+    snow_data = None
+    snow_band = None
+    snow_tiff = None
+    masked_hfi_tiff_band = None
+    masked_hfi_tiff = None
+
+    return masked_hfi_path
+
+
+async def apply_snow_mask(hfi_path: str, last_processed_snow: ProcessedSnow, temp_dir: str):
+    gdal.SetConfigOption('AWS_SECRET_ACCESS_KEY', config.get('OBJECT_STORE_SECRET'))
+    gdal.SetConfigOption('AWS_ACCESS_KEY_ID', config.get('OBJECT_STORE_USER_ID'))
+    gdal.SetConfigOption('AWS_S3_ENDPOINT', config.get('OBJECT_STORE_SERVER'))
+    gdal.SetConfigOption('AWS_VIRTUAL_HOSTING', 'FALSE')
+
+    snow_mask_path = await prepare_snow_mask(hfi_path, last_processed_snow, temp_dir)
```

Comment:
Any idea if the `temp_dir` is given a random / hashed name to avoid conflicts? I'm just wondering what would happen if 2 or more of these processing jobs get run at the same time on the same machine.

Reply:
I've never looked into it before now! The name of the `tempfile.TemporaryDirectory` is generated randomly. Here's the docstring from the underlying method.

`class _RandomNameSequence:
    """An instance of _RandomNameSequence generates an endless
    sequence of unpredictable strings which can safely be incorporated
    into file names.  Each string is eight characters long.  Multiple
    threads can safely use the same instance at the same time.

    _RandomNameSequence is an iterator."""
`